### DISCUSSION TITLE
Does Diesel support "dynamically-named tables" in its schema?

### DISCUSSION THREAD

**Discussion by AllianceDarkSylinc** [diesel-rs/diesel#3860] [Q&A]
Hi!

I'm learning Diesel to use as a potential replalcement for our current backend (videogame server) and we support stuff like the following:

```mysql
CREATE TABLE user_accounts( /*....*/ );

CREATE TABLE {$game_name}_score( /*....*/ );

// For example:
CREATE TABLE tetris_score( /*....*/ );
CREATE TABLE pacman_score( /*....*/ );
```

In this example, 'user_accounts' is shared by all games the server supports. Functionally 'tetris_score' & 'pacman_score' are exactly the same. They just contain data belonging to different games.

The code that handles this from rust should be the same (otherwise it would be lots of copy-paste code). The name(s) of the table(s) the runtime will support _can_ be known at compile-time though.

The current backend is not ORM and constructs the table names at runtime while forming the query. It first validates the game names, generates the query string and executes the prepared statement.

My guess the answer will be "no, diesel is not meant to work this way", but I'm asking just in case to see if something like this can be support via Diesel (via ORM of course).

## Update

I'm new to Rust as well. What I'm asking is that perhaps it is possible to do something like this (since game names can be known at compile time):

```rust
#[derive(Queryable, Selectable, Insertable)]
#[diesel(table_name = T_scores)]
pub struct Score<T> {
     // ..data..
};

let score = Score<pacman>();
```

**Comment by weiznich** [comment]
It's possible to use diesel with schemas not know at compile time via the [`diesel_dynamic_schema`](https://docs.diesel.rs/2.1.x/diesel_dynamic_schema/index.html) crate. You should be able to implement your usecase using the methods provided by that crate. 

As for derives: It depends on the derive, if you can use them or not
* `Queryable` is likely usable, as it doesn't need to know anything about your table. It only cares about query return types matching your struct
* `Insertable` and `Selectable` are likely not usable as they depend on the static table definition.

The alternative is to restructure your database schema to have a more classical layout. That would mean having only one `scores` table which contains all scores for all games. Games are then identified via a separate table and linked by a foreign key to the `scores` table.
